### PR TITLE
Enable ManagedMediaSource compilation on GTK platforms

### PR DESCRIPTION
**Pull request by jyavenard**
#### 86c12d6b8e31bfe980de1d28cddd43d623f2f6d2
<pre>
Enable ManagedMediaSource compilation on GTK platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=252386">https://bugs.webkit.org/show_bug.cgi?id=252386</a>
rdar://105540966

Reviewed by Philippe Normand.

Copy ENABLE_MEDIA_SOURCE and related variables to enable conditional
compilation of ManagedMediaSource.
Code is made dependent on ENABLE_MEDIA_SOURCE and ENABLE_MANAGED_MEDIA_SOURCE
to be both set 1.

* LayoutTests/media/media-source/media-managedmse-expected.txt:
* LayoutTests/media/media-source/media-managedmse.html:
* LayoutTests/platform/glib/TestExpectations: enable ManagedMediaSource test.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/cmake/GStreamerDefinitions.cmake:
* Source/cmake/OptionsMac.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Source/cmake/tools/vsprops/FeatureDefines.props:
* Source/cmake/tools/vsprops/FeatureDefinesCairo.props:
* Tools/Scripts/webkitperl/FeatureList.pm:

Canonical link: <a href="https://commits.webkit.org/260671@main">https://commits.webkit.org/260671@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/032c0134e0a712d4efb25c87d1def51bac0c476d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109021 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18100 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/557 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118296 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112904 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9401 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101259 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114779 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/97900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42794 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29543 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10885 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30891 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8983 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11629 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17009 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50489 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106522 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7368 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13228 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | | | |

### CODE REVIEW

**Comment by philn** [comment]
I think changes are needed in OptionsWPE.cmake and OptionsGTK.cmake as well.

File: Source/cmake/GStreamerDefinitions.cmake [hunk 1]
```diff
@@ -1,6 +1,7 @@
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_VIDEO PUBLIC ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEB_AUDIO PUBLIC ON)
 
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MANAGED_MEDIA_SOURCE PUBLIC ON)
```
Comment (mcatanzaro) [inline Source/cmake/GStreamerDefinitions.cmake]: I don't think this option should be public unless you have a very strong reason for people to disable it...

Comment (jyavenard) [reply Source/cmake/GStreamerDefinitions.cmake]: We want this option to be disabled by default for now. But compiled in so it can be enabled by a flag.

Ultimately, it's all very similar to ENABLE_MEDIA_SOURCE ; why would that one gets an option but not the other?

**Comment by jyavenard** [comment]
> I think changes are needed in OptionsWPE.cmake and OptionsGTK.cmake as well.

None of those two had ENABLE_MEDIA_SOURCE ; are they missing and should they be there?

**Comment by jyavenard** [comment]
> > I think changes are needed in OptionsWPE.cmake and OptionsGTK.cmake as well.
> 
> None of those two had ENABLE_MEDIA_SOURCE ; are they missing and should they be there?

Actually, from the documentation on those files: 

```# Private options shared with other WebKit ports. Add options here when
# we need a value different from the default defined in WebKitFeatures.cmake.
# Changing these options is completely unsupported.
```

and those are defined in WebKitFeatures.cmake if ENABLE_VIDEO is set.

File: Source/cmake/GStreamerDefinitions.cmake (hunk 1 above)
Comment (jyavenard) [reply Source/cmake/GStreamerDefinitions.cmake]: I've removed this line.

File: Tools/Scripts/webkitperl/FeatureList.pm [hunk 2]
```diff
@@ -379,6 +380,9 @@ my @features = (
     { option => "media-controls-script", desc => "Toggle definition of media controls in Javascript",
       define => "ENABLE_MEDIA_CONTROLS_SCRIPT", value => \$mediaControlsScriptSupport },
 
+    { option => "managed-media-source", desc => "Toggle Managed Media Source support",
+      define => "ENABLE_MANAGED_MEDIA_SOURCE", value => \$managedMediaSourceSupport },
```
Comment (mcatanzaro) [inline Tools/Scripts/webkitperl/FeatureList.pm]: Please alphabetize it

File: Source/cmake/OptionsWin.cmake [hunk 3]
```diff
@@ -142,6 +142,7 @@ endif ()
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_LAYOUT_TESTS PUBLIC ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MATHML PUBLIC ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MEDIA_CONTROLS_SCRIPT PUBLIC ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MANAGED_MEDIA_SOURCE PUBLIC OFF)
```
Comment (mcatanzaro) [inline Source/cmake/OptionsWin.cmake]: No need for this: it's redundant with the default value in WebKitFeatures.cmake. (And not alphabetized.)

File: Source/cmake/GStreamerDefinitions.cmake (hunk 1 above)
Comment (mcatanzaro) [reply Source/cmake/GStreamerDefinitions.cmake]: I think ENABLE_MEDIA_SOURCE is public because it originally required a newer version of GStreamer. Since that's no longer true, it no longer needs to be public (unless somebody has a real good use case for why MSE would need to be disabled at build time).

**Comment by mcatanzaro** [comment]
This looks good now.

File: Source/cmake/OptionsWin.cmake (hunk 3 above)
Comment (jyavenard) [reply Source/cmake/OptionsWin.cmake]: isn't it redundant for ENABLE_MEDIA_SOURCE too ?

Comment (jyavenard) [reply Source/cmake/OptionsWin.cmake]: Actually, it's not the same as the default value in WebKitFeatures.cmake which is based on the value of ENABLE_VIDEO ; here it's turned off.
But regardless all the code is dependent on ENABLE(MANAGED_MEDIA_SOURCE) && ENABLE(MEDIA_SOURCE) so if ENABLE_MEDIA_SOURCE is off as it is here, it will be disabled as well in the code.

File: Tools/Scripts/webkitperl/FeatureList.pm (hunk 2 above)
Comment (jyavenard) [reply Tools/Scripts/webkitperl/FeatureList.pm]: done

File: Source/cmake/GStreamerDefinitions.cmake (hunk 1 above)
Comment (jyavenard) [reply Source/cmake/GStreamerDefinitions.cmake]: Actually, this is the only place enabling that value (both ENABLE_MEDIA_SOURCE and ENABLE_MANAGED_MEDIA_SOURCE) so it's definitely required.

Would PRIVATE be more suitable here (I have no clue what those are for)

Comment (mcatanzaro) [reply Source/cmake/GStreamerDefinitions.cmake]: PRIVATE is equivalent to CMake ADVANCED. It means the option is not exposed to end users (unless they peek behind the curtain). You need a very strong reason for it to be PUBLIC and I don't see one here, so it should be PRIVATE unless you have some justification for PUBLIC.

> Actually, this is the only place enabling that value (both ENABLE_MEDIA_SOURCE and ENABLE_MANAGED_MEDIA_SOURCE) so it's definitely required.

Well you originally said you want it disabled by default. If so, it must be OFF. Turning it ON means it's enabled by default. You can only choose one way or the other... which is it that you want? :P

File: Source/cmake/WebKitFeatures.cmake [hunk 4]
```diff
@@ -285,6 +286,7 @@ macro(WEBKIT_OPTION_BEGIN)
     WEBKIT_OPTION_DEPEND(ENABLE_MEDIA_CONTROLS_CONTEXT_MENUS ENABLE_MODERN_MEDIA_CONTROLS)
     WEBKIT_OPTION_DEPEND(ENABLE_MEDIA_RECORDER ENABLE_MEDIA_STREAM)
     WEBKIT_OPTION_DEPEND(ENABLE_MEDIA_SESSION ENABLE_VIDEO)
+    WEBKIT_OPTION_DEPEND(ENABLE_MANAGED_MEDIA_SOURCE ENABLE_MEDIA_SESSION)
```
Comment (mcatanzaro) [inline Source/cmake/WebKitFeatures.cmake]: Still needs to be alphabetized.

Did you mean for it to depend on `ENABLE_MEDIA_SOURCE` instead, or in addition to, `ENABLE_MEDIA_SESSION`?

File: Source/cmake/OptionsWin.cmake (hunk 3 above)
Comment (mcatanzaro) [reply Source/cmake/OptionsWin.cmake]: You have in WebKitFeatures.cmake:

` WEBKIT_OPTION_DEFINE(ENABLE_MANAGED_MEDIA_SOURCE "Toggle Managed Media Source support" PRIVATE OFF)`

So it's OFF by default. The dependencies will ensure the feature gets unset if required, but the dependencies won't ever enable the feature.

File: LayoutTests/platform/glib/TestExpectations [hunk 5]
```diff
@@ -1017,8 +1017,6 @@ webkit.org/b/211995 fast/images/animated-image-mp4.html [ Failure Timeout ]
 
 webkit.org/b/214038 imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-cues-cuechange.html [ Missing Failure ]
 
-webkit.org/b/252386 media/media-source/media-managedmse.html [ Failure ]
```
Comment (mcatanzaro) [inline LayoutTests/platform/glib/TestExpectations:1021]: It seems the test is failing. :(

Comment (jyavenard) [reply LayoutTests/platform/glib/TestExpectations:1021]: Failing it seems because gstreamer demux the webm content with different timestamp to libwebm. 

At least the code is compiled now

File: Source/cmake/WebKitFeatures.cmake (hunk 4 above)
Comment (jyavenard) [reply Source/cmake/WebKitFeatures.cmake]: Most in this file aren't alphabetised 
```
    WEBKIT_OPTION_DEPEND(ENABLE_ACCESSIBILITY_ISOLATED_TREE ENABLE_ACCESSIBILITY)
    WEBKIT_OPTION_DEPEND(ENABLE_WEB_RTC ENABLE_MEDIA_STREAM)
    WEBKIT_OPTION_DEPEND(ENABLE_ENCRYPTED_MEDIA ENABLE_VIDEO)
    WEBKIT_OPTION_DEPEND(ENABLE_LEGACY_ENCRYPTED_MEDIA ENABLE_VIDEO)
    WEBKIT_OPTION_DEPEND(ENABLE_DFG_JIT ENABLE_JIT)
    WEBKIT_OPTION_DEPEND(ENABLE_FTL_JIT ENABLE_DFG_JIT)
    WEBKIT_OPTION_DEPEND(ENABLE_WEBASSEMBLY ENABLE_JIT)
    WEBKIT_OPTION_DEPEND(ENABLE_WEBASSEMBLY_B3JIT ENABLE_FTL_JIT)
    WEBKIT_OPTION_DEPEND(ENABLE_INSPECTOR_ALTERNATE_DISPATCHERS ENABLE_REMOTE_INSPECTOR)
 ```
Couldn't be more out of order.
 As such, I think it makes more sense to keep MANAGED_MEDIA_SOURCE right next to MEDIA_SOURCE seeing they are extremely closely related.

**Comment by philn** [comment]
Dunno what's up with this EWS, but the new test runs as expected here. Feel free to ignore the red flag.

**Comment by mcatanzaro** [comment]
> Dunno what's up with this EWS, but the new test runs as expected here. Feel free to ignore the red flag.

```
--- /home/ews/worker/GTK-WK2-Tests-EWS/build/layout-test-results/media/media-source/media-managedmse-expected.txt
+++ /home/ews/worker/GTK-WK2-Tests-EWS/build/layout-test-results/media/media-source/media-managedmse-actual.txt
@@ -1,17 +1,4 @@
 
-RUN(video.src = URL.createObjectURL(source))
-EVENT(sourceopen)
-RUN(sourceBuffer = source.addSourceBuffer(loader.type()))
-RUN(sourceBuffer.appendBuffer(loader.initSegment()))
-EVENT(update)
-Append a media segment.
-RUN(sourceBuffer.appendBuffer(loader.mediaSegment(0)))
-EVENT(update)
-EXPECTED (sourceBuffer.buffered.length == '1') OK
-EXPECTED (Math.abs(sourceBuffer.buffered.end(0) - loader.mediaSegmentEndTime(0)) < 0.01 == 'true') OK
-Clean sourcebuffer of all content.
-RUN(sourceBuffer.remove(0, 100))
-EVENT(update)
-EXPECTED (sourceBuffer.buffered.length == '0') OK
+Caught exception: "ReferenceError: Can't find variable: ManagedMediaSource" FAIL
 END OF TEST
```

Looks like the feature is not actually enabled.

**Comment by mcatanzaro** [comment]
The feature changes look fine to me now, although I don't understand why the test is failing.

Note that you have enabled the new feature by default. That seems fine to me, but I wanted to make sure it's intentional because previously you said you did not want to do that.

**Comment by jyavenard** [comment]
> The feature changes look fine to me now, although I don't understand why the test is failing.
> 
> Note that you have enabled the new feature by default. That seems fine to me, but I wanted to make sure it's intentional because previously you said you did not want to do that.

What I did is make ENABLE_MANAGED_MEDIA_SOURCE be 1 by default, but now everything depends on `(ENABLE_MANAGED_MEDIA_SOURCE && ENABLE_MEDIA_SOURCE)` rather than just `ENABLE_MANAGED_MEDIA_SOURCE` ; I thought it would make things simpler.

I don't understand what is happening with this EWS build, I can get it to work if I add `WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MANAGED_MEDIA_SOURCE PUBLIC ON)` to GStreamerDefinitions.cmake

which makes no sense.

It works fine on my local ubuntu machine and gtk build

**Review by philn — APPROVED** [review]

File: Source/cmake/GStreamerDefinitions.cmake [hunk 6]
```diff
@@ -2,6 +2,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_VIDEO PUBLIC ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEB_AUDIO PUBLIC ON)
 
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MEDIA_SOURCE PUBLIC ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MANAGED_MEDIA_SOURCE PUBLIC ON)
```
Comment (mcatanzaro) [inline Source/cmake/GStreamerDefinitions.cmake]: The new option needs to be PRIVATE, and might as well drive-by change the existing option too, because there's no longer any justifiable reason for it to be PUBLIC:

```suggestion
WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MEDIA_SOURCE PRIVATE ON)
WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MANAGED_MEDIA_SOURCE PRIVATE ON)
```

**Review by mcatanzaro — COMMENTED** [review]
OK so your intention is to enable it by default at build time but not runtime? That's fine.
